src/components/TeacherComponents/OnboardingForm/OnboardingForm.schema.ts

```diff
@@ -4,15 +4,17 @@ const baseSchema = z.object({
   newsletterSignUp: z.boolean(),
 });
 
-export const roleSelectFormSchema = z.object({
-  ...baseSchema.shape,
-  role: z.string({
-    errorMap: () => ({
-      message: "Select a role",
-    }),
-  }),
-  other: z.string().optional(),
-});
+export const roleSelectFormSchema = z
+  .object({
+    ...baseSchema.shape,
+    role: z.string({ message: "Please select what describes you best" }),
+    other: z.string().trim().optional(),
+  })
+  .refine((input) => input.role === "Other" && !!input.other, {
```

Comment:
oh this is neat! didn't know zod could do this

Reply:
Was new to me too! Though I wish this sort of inter relationship could be expressed declaratively and a more representative type inferred as a result. Right now the type doesn't respect the rules in the refine so `other` will always be optional.

I've never used it but [Yup has `when` ](https://github.com/jquense/yup?tab=readme-ov-file#schemawhenkeys-string--string-builder-object--values-any-schema--schema-schema) which is a bit more declarative and does the same job — though i'm not sure if it will infer a better type!